token/core/zkatdlog/crypto/validator/validator.go

```diff
@@ -8,31 +8,99 @@ package validator
 
 import (
 	"bytes"
+	"encoding/json"
+	"time"
 
 	math "github.com/IBM/mathlib"
 	"github.com/hyperledger-labs/fabric-smart-client/platform/view/services/flogging"
 	"github.com/hyperledger-labs/fabric-smart-client/platform/view/services/hash"
 	"github.com/hyperledger-labs/fabric-smart-client/platform/view/view"
-	"github.com/pkg/errors"
-
+	"github.com/hyperledger-labs/fabric-token-sdk/token/core/identity"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/core/interop"
 	"github.com/hyperledger-labs/fabric-token-sdk/token/core/zkatdlog/crypto"
 	issue2 "github.com/hyperledger-labs/fabric-token-sdk/token/core/zkatdlog/crypto/issue"
 	"github.com/hyperledger-labs/fabric-token-sdk/token/core/zkatdlog/crypto/token"
 	"github.com/hyperledger-labs/fabric-token-sdk/token/core/zkatdlog/crypto/transfer"
 	"github.com/hyperledger-labs/fabric-token-sdk/token/driver"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/interop/exchange"
+	"github.com/pkg/errors"
 )
 
 var logger = flogging.MustGetLogger("token-sdk.zkatdlog")
 
+var defaultValidators = []ValidateTransfer{SerializedIdentityTypeExtraValidator, ScriptTypeExchangeExtraValidator}
+
+type ValidateTransfer func(tokens []*token.Token, tr *transfer.TransferAction) error
+
+func SerializedIdentityTypeExtraValidator(tokens []*token.Token, tr *transfer.TransferAction) error {
+	// noting else to validate
+	return nil
+}
+
+func ScriptTypeExchangeExtraValidator(tokens []*token.Token, tr *transfer.TransferAction) error {
+	for _, in := range tokens {
+		owner, err := identity.UnmarshallRawOwner(in.Owner)
+		if err != nil {
+			return errors.Wrap(err, "failed to unmarshal owner of input token")
+		}
+		if owner.Type == exchange.ScriptTypeExchange {
+			if len(tokens) != 1 || len(tr.GetOutputs()) != 1 {
+				return errors.Errorf("invalid transfer action: an exchange script only transfers the ownership of a token")
+			}
+
+			out := tr.GetOutputs()[0].(*token.Token)
+			if tokens[0].Data.Equals(out.Data) {
+				return errors.Errorf("invalid transfer action: content of input does not match content of output")
+			}
+
+			// check that owner field in output is correct
+			if err := interop.VerifyTransferFromExchangeScript(tokens[0].Owner, out.Owner); err != nil {
+				return errors.Wrap(err, "failed to verify transfer from exchange script")
+			}
+		}
+	}
+
+	for _, o := range tr.GetOutputs() {
```

Comment:
why is redeem checked here?

Reply:
Just to make sure we don't create an error when trying to unmarshall later